File: backend/src/socket/socketHandler.js
import Message from '../models/Message.js';
import Conversation from '../models/Conversation.js';
import { createNotification, createNotifications } from '../utils/notificationService.js';

export const setupSocket = (io) => {
    let onlineUsers = new Map();
    const videoRooms = new Map();
    const ringingCalls = new Map();

    io.on('connection', (socket) => {
        socket.emit("me", socket.id);

        socket.on("setup", (userData) => {
            const userId = userData?._id || userData?.id;
            if (userId) {
                socket.join(userId);
                onlineUsers.set(userId, socket.id);
                if (userData.role === 'master_admin') {
                    socket.join('platform_admin');
                }
                io.emit("onlineUsers", Array.from(onlineUsers.keys()));
                socket.emit("connected");
            }
        });

        socket.on("setup_dashboard", ({ teamIds }) => {
            if (Array.from(teamIds).length > 0) {
                teamIds.forEach(id => {
                    socket.join(`team_${id}`);
                });
            }
        });

        socket.on("disconnect", () => {
            let disconnectedUserId;
            for (let [userId, socketId] of onlineUsers.entries()) {
                if (socketId === socket.id) {
                    disconnectedUserId = userId;
                    break;
                }
            }

            if (disconnectedUserId) {
                onlineUsers.delete(disconnectedUserId);
                io.emit("onlineUsers", Array.from(onlineUsers.keys()));
            }

            if (ringingCalls.has(socket.id)) {
                ringingCalls.delete(socket.id);
            }
            
            if (disconnectedUserId) {
                for (let [callerSocketId, callInfo] of ringingCalls.entries()) {
                    if (callInfo.recipientId === disconnectedUserId) {
                        ringingCalls.delete(callerSocketId);
                    }
                }
            }

            videoRooms.forEach((room, roomId) => {
                room.forEach((participant, userId) => {
                    if (participant.socketId === socket.id) {
                        room.delete(userId);
                        const participants = Array.from(room.values()).map(p => ({
                            id: p.userId,
                            name: p.userName
                        }));
                        io.to(roomId).emit("userLeftRoom", {
                            userId,
                            participants
                        });
                        if (room.size === 0) {
                            videoRooms.delete(roomId);
                        }
                    }
                });
            });
        });

        socket.on("joinRoom", ({ roomId, userId, userName, name }) => {
            if (userId && userName) {
                if (!videoRooms.has(roomId)) {
                    videoRooms.set(roomId, new Map());
                }
                const room = videoRooms.get(roomId);
                const uid = userId.toString();
                room.set(uid, { userId: uid, userName, socketId: socket.id });
                socket.join(roomId);
                const participants = Array.from(room.values()).map(p => ({
                    id: p.userId,
                    name: p.userName
                }));
                socket.emit("roomJoined", { participants });
                socket.to(roomId).emit("userJoinedRoom", {
                    userId: uid,
                    userName,
                    participants
                });
            } else {
                socket.join(roomId);
                socket.to(roomId).emit("userJoined", { id: socket.id, name });
            }
        });

        socket.on("leaveRoom", ({ roomId, userId }) => {
            const room = videoRooms.get(roomId);
            if (room) {
                room.delete(userId);
                socket.leave(roomId);
                const participants = Array.from(room.values()).map(p => ({
                    id: p.userId,
                    name: p.userName
                }));
                io.to(roomId).emit("userLeftRoom", {
                    userId,
                    participants
                });
                if (room.size === 0) {
                    videoRooms.delete(roomId);
                }
            }
        });

        socket.on("sendOffer", ({ to, offer, roomId }) => {
            const room = videoRooms.get(roomId);
            if (room) {
                const recipient = Array.from(room.values()).find(p => p.userId === to);
                if (recipient) {
                    const sender = Array.from(room.values()).find(p => p.socketId === socket.id);
                    if (sender) {
                        io.to(recipient.socketId).emit("receiveOffer", {
                            from: sender.userId,
                            fromName: sender.userName,
                            offer
                        });
                    }
                }
            }
        });

        socket.on("sendAnswer", ({ to, answer, roomId }) => {
            const room = videoRooms.get(roomId);
            if (room) {
                const recipient = Array.from(room.values()).find(p => p.userId === to);
                if (recipient) {
                    const sender = Array.from(room.values()).find(p => p.socketId === socket.id);
                    if (sender) {
                        io.to(recipient.socketId).emit("receiveAnswer", {
                            from: sender.userId,
                            answer
                        });
                    }
                }
            }
        });

        socket.on("sendIceCandidate", ({ to, candidate, roomId }) => {
            const room = videoRooms.get(roomId);
            if (room) {
                const recipient = Array.from(room.values()).find(p => p.userId === to);
                if (recipient) {
                    io.to(recipient.socketId).emit("receiveIceCandidate", {
                        from: Array.from(room.values()).find(p => p.socketId === socket.id)?.userId,
                        candidate
                    });
                }
            }
        });

        socket.on("roomMessage", ({ roomId, message }) => {
            socket.to(roomId).emit("roomMessage", { message });
        });

        socket.on("callUser", async ({ userToCall, signalData, from, name, isVideo }) => {
            ringingCalls.set(socket.id, {
                recipientId: userToCall,
                callerId: from,
                callerName: name,
                isVideo: !!isVideo
            });
            io.to(userToCall).emit("callUser", { signal: signalData, from, name, isVideo });

            // Notify master_admin about the ongoing call (without sending the call signal to them)
            try {
                await createNotifications([userToCall], {
                    sender: from,
                    title: `Incoming ${isVideo ? 'Video' : 'Voice'} Call`,
                    description: `${name} is calling${isVideo ? ' (video)' : ''}`,
                    type: 'message',
                    link: '/chat'
                }, io);
            } catch (err) {
                console.error('[Socket] callUser notification error:', err);
            }
        });

        socket.on("answerCall", (data) => {
            for (let [callerSocketId, callInfo] of ringingCalls.entries()) {
                if (callInfo.recipientId === data.to || callInfo.recipientId === socket.id) {
                     ringingCalls.delete(callerSocketId);
                }
            }
            io.to(data.to).emit("callAccepted", data.signal);
        });

        socket.on("ice-candidate", ({ to, candidate }) => {
            io.to(to).emit("ice-candidate", candidate);
        });

        socket.on("endCall", async ({ to }) => {
            const callInfo = ringingCalls.get(socket.id);
            if (callInfo && callInfo.recipientId === to) {
                try {
                    await createNotification(to, {
                        sender: callInfo.callerId,
                        title: 'Missed Call',
                        description: `You missed a ${callInfo.isVideo ? 'video' : 'voice'} call from ${callInfo.callerName}`,
                        type: 'message',
                        link: '/chat'
                    }, io);
                } catch (error) {
                    console.error("Error creating missed call notification:", error);
                }
                ringingCalls.delete(socket.id);
            }
            io.to(to).emit("callEnded");
        });

        socket.on("joinChat", (chatId) => {
            if (!chatId) return;
            socket.join(chatId);
        });

        socket.on("leaveChat", (chatId) => {
            if (!chatId) return;
            socket.leave(chatId);
        });

        socket.on("typing", (room) => socket.in(room).emit("typing", room));
        socket.on("stopTyping", (room) => socket.in(room).emit("stopTyping", room));

        socket.on("sendMessage", async ({ room, messageId, text, senderId, senderName, type }) => {
            try {
                const conversation = await Conversation.findById(room)
                    .populate('users', '_id name email');

                if (!conversation) return;

                const broadcastPayload = {
                    _id: messageId,
                    chat: {
                        _id: conversation._id,
                        chatName: conversation.chatName,
                        isGroupChat: conversation.isGroupChat,
                        users: conversation.users,
                        groupAdmin: conversation.groupAdmin
                    },
                    text,
                    sender: { _id: senderId, name: senderName },
                    type: type || 'text',
                    createdAt: new Date().toISOString()
                };

                socket.to(room).emit("receiveMessage", broadcastPayload);

                // Collect direct participant IDs (excluding sender)
                const recipientIds = conversation.users
                    .filter(u => u._id.toString() !== senderId.toString())
                    .map(u => u._id.toString());

                recipientIds.forEach(uid => {
                    io.to(uid).emit("receiveMessage", broadcastPayload);
                });

                // Create DB notification for recipients + auto-include master_admins
                // This ensures master_admin gets a newNotification event even if not in the chat
                if (recipientIds.length > 0) {
                    const chatLabel = conversation.isGroupChat
                        ? conversation.chatName
                        : senderName;
                    const preview = text && text.length > 80
                        ? text.substring(0, 77) + '...'
                        : (text || (type === 'file' ? '📎 File shared' : 'New message'));

                    await createNotifications(recipientIds, {
                        sender: senderId,
                        title: `New message from ${chatLabel}`,
                        description: preview,
                        type: 'message',
                        link: '/chat'
                    }, io);
                }
            } catch (error) {
                console.error("Socket sendMessage error:", error);
            }
        });
    });
};

File: backend/socket/socket_handler.go
package socket

import (
	"log"
	"net/http"
	"sync"

	"github.com/zishang520/engine.io/v2/types"
	"github.com/zishang520/socket.io/v2/socket"
)

type Participant struct {
	UserID   string `json:"id"`
	UserName string `json:"name"`
	SocketID string `json:"-"`
}

type RingingCall struct {
	RecipientID string `json:"recipientId"`
	CallerID    string `json:"callerId"`
	CallerName  string `json:"callerName"`
	IsVideo     bool   `json:"isVideo"`
}

type SocketManager struct {
	mu           sync.RWMutex
	Server       *socket.Server
	OnlineUsers  map[string]socket.SocketId         // userId -> socketId
	VideoRooms   map[string]map[string]*Participant // roomId -> map[userId]*Participant
	RingingCalls map[socket.SocketId]*RingingCall   // callerSocketId -> RingingCall
}

var Manager *SocketManager

// InitSocketServer initializes the Socket.IO v4 server and handlers
func InitSocketServer() (*socket.Server, http.Handler) {
	opts := socket.DefaultServerOptions()
	opts.SetAllowEIO3(true)
	opts.SetCors(&types.Cors{
		Origin:      true,
		Credentials: true,
	})
	opts.SetTransports(types.NewSet("polling", "websocket"))

	io := socket.NewServer(nil, opts)

	Manager = &SocketManager{
		Server:       io,
		OnlineUsers:  make(map[string]socket.SocketId),
		VideoRooms:   make(map[string]map[string]*Participant),
		RingingCalls: make(map[socket.SocketId]*RingingCall),
	}

	io.On("connection", func(clients ...any) {
		client := clients[0].(*socket.Socket)
		socketID := client.Id()

		// Send "me" event
		client.Emit("me", string(socketID))

		// setup event: user logging in or refreshing
		client.On("setup", func(args ...any) {
			if len(args) == 0 {
				return
			}
			userData, ok := args[0].(map[string]any)
			if !ok {
				return
			}

			var userID string
			if id, ok := userData["_id"].(string); ok && id != "" {
				userID = id
			} else if id, ok := userData["id"].(string); ok && id != "" {
				userID = id
			}

			if userID != "" {
				client.Join(socket.Room(userID))

				Manager.mu.Lock()
				Manager.OnlineUsers[userID] = socketID
				if role, ok := userData["role"].(string); ok && role == "master_admin" {
					client.Join("platform_admin")
				}
				onlineKeys := make([]string, 0, len(Manager.OnlineUsers))
				for k := range Manager.OnlineUsers {
					onlineKeys = append(onlineKeys, k)
				}
				Manager.mu.Unlock()

				io.Emit("onlineUsers", onlineKeys)
				client.Emit("connected")
			}
		})

		// setup_dashboard: join team rooms
		client.On("setup_dashboard", func(args ...any) {
			if len(args) == 0 {
				return
			}
			payload, ok := args[0].(map[string]any)
			if !ok {
				return
			}
			if teamIDs, ok := payload["teamIds"].([]any); ok {
				for _, tid := range teamIDs {
					if idStr, ok := tid.(string); ok {
						client.Join(socket.Room("team_" + idStr))
					}
				}
			}
		})

		// joinRoom: video call room
		client.On("joinRoom", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, ok := args[0].(map[string]any)
			if !ok {
				return
			}

			roomID, _ := data["roomId"].(string)
			userID, _ := data["userId"].(string)
			userName, _ := data["userName"].(string)
			name, _ := data["name"].(string)

			if userName == "" && name != "" {
				userName = name
			}

			if roomID != "" && userID != "" && userName != "" {
				client.Join(socket.Room(roomID))

				Manager.mu.Lock()
				room, exists := Manager.VideoRooms[roomID]
				if !exists {
					room = make(map[string]*Participant)
					Manager.VideoRooms[roomID] = room
				}
				room[userID] = &Participant{
					UserID:   userID,
					UserName: userName,
					SocketID: string(socketID),
				}

				participants := make([]map[string]string, 0, len(room))
				for _, p := range room {
					participants = append(participants, map[string]string{
						"id":   p.UserID,
						"name": p.UserName,
					})
				}
				Manager.mu.Unlock()

				client.Emit("roomJoined", map[string]any{"participants": participants})
				client.To(socket.Room(roomID)).Emit("userJoinedRoom", map[string]any{
					"userId":       userID,
					"userName":     userName,
					"participants": participants,
				})
			}
		})

		// leaveRoom: video call room
		client.On("leaveRoom", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, ok := args[0].(map[string]any)
			if !ok {
				return
			}
			roomID, _ := data["roomId"].(string)
			userID, _ := data["userId"].(string)

			if roomID != "" {
				Manager.mu.Lock()
				if room, ok := Manager.VideoRooms[roomID]; ok {
					delete(room, userID)
					client.Leave(socket.Room(roomID))

					participants := make([]map[string]string, 0, len(room))
					for _, p := range room {
						participants = append(participants, map[string]string{
							"id":   p.UserID,
							"name": p.UserName,
						})
					}
					if len(room) == 0 {
						delete(Manager.VideoRooms, roomID)
					}
					Manager.mu.Unlock()

					io.To(socket.Room(roomID)).Emit("userLeftRoom", map[string]any{
						"userId":       userID,
						"participants": participants,
					})
				} else {
					Manager.mu.Unlock()
				}
			}
		})

		// WebRTC offer / answer / ice
		client.On("sendOffer", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			to, _ := data["to"].(string)
			offer := data["offer"]
			roomID, _ := data["roomId"].(string)

			Manager.mu.RLock()
			room := Manager.VideoRooms[roomID]
			var recipientSocketID socket.SocketId
			var senderID, senderName string

			if room != nil {
				if r, ok := room[to]; ok {
					recipientSocketID = socket.SocketId(r.SocketID)
				}
				for uid, p := range room {
					if p.SocketID == string(socketID) {
						senderID = uid
						senderName = p.UserName
						break
					}
				}
			}
			Manager.mu.RUnlock()

			if recipientSocketID != "" {
				io.To(socket.Room(recipientSocketID)).Emit("receiveOffer", map[string]any{
					"from":     senderID,
					"fromName": senderName,
					"offer":    offer,
				})
			}
		})

		client.On("sendAnswer", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			to, _ := data["to"].(string)
			answer := data["answer"]
			roomID, _ := data["roomId"].(string)

			Manager.mu.RLock()
			room := Manager.VideoRooms[roomID]
			var recipientSocketID socket.SocketId
			var senderID string

			if room != nil {
				if r, ok := room[to]; ok {
					recipientSocketID = socket.SocketId(r.SocketID)
				}
				for uid, p := range room {
					if p.SocketID == string(socketID) {
						senderID = uid
						break
					}
				}
			}
			Manager.mu.RUnlock()

			if recipientSocketID != "" {
				io.To(socket.Room(recipientSocketID)).Emit("receiveAnswer", map[string]any{
					"from":   senderID,
					"answer": answer,
				})
			}
		})

		client.On("sendIceCandidate", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			to, _ := data["to"].(string)
			candidate := data["candidate"]
			roomID, _ := data["roomId"].(string)

			Manager.mu.RLock()
			room := Manager.VideoRooms[roomID]
			var recipientSocketID socket.SocketId
			var senderID string

			if room != nil {
				if r, ok := room[to]; ok {
					recipientSocketID = socket.SocketId(r.SocketID)
				}
				for uid, p := range room {
					if p.SocketID == string(socketID) {
						senderID = uid
						break
					}
				}
			}
			Manager.mu.RUnlock()

			if recipientSocketID != "" {
				io.To(socket.Room(recipientSocketID)).Emit("receiveIceCandidate", map[string]any{
					"from":      senderID,
					"candidate": candidate,
				})
			}
		})

		client.On("roomMessage", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			roomID, _ := data["roomId"].(string)
			msg := data["message"]

			if roomID != "" {
				client.To(socket.Room(roomID)).Emit("roomMessage", map[string]any{"message": msg})
			}
		})

		// 1-on-1 Calling
		client.On("callUser", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			userToCall, _ := data["userToCall"].(string)
			signalData := data["signalData"]
			from, _ := data["from"].(string)
			name, _ := data["name"].(string)
			isVideo, _ := data["isVideo"].(bool)

			Manager.mu.Lock()
			Manager.RingingCalls[socketID] = &RingingCall{
				RecipientID: userToCall,
				CallerID:    from,
				CallerName:  name,
				IsVideo:     isVideo,
			}
			Manager.mu.Unlock()

			io.To(socket.Room(userToCall)).Emit("callUser", map[string]any{
				"signal":  signalData,
				"from":    from,
				"name":    name,
				"isVideo": isVideo,
			})
		})

		client.On("answerCall", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			to, _ := data["to"].(string)
			signal := data["signal"]

			Manager.mu.Lock()
			for callerID, info := range Manager.RingingCalls {
				if info.RecipientID == to || string(callerID) == to {
					delete(Manager.RingingCalls, callerID)
				}
			}
			Manager.mu.Unlock()

			io.To(socket.Room(to)).Emit("callAccepted", signal)
		})

		client.On("ice-candidate", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			to, _ := data["to"].(string)
			cand := data["candidate"]
			io.To(socket.Room(to)).Emit("ice-candidate", cand)
		})

		client.On("endCall", func(args ...any) {
			if len(args) == 0 {
				return
			}
			data, _ := args[0].(map[string]any)
			to, _ := data["to"].(string)

			Manager.mu.Lock()
			delete(Manager.RingingCalls, socketID)
			Manager.mu.Unlock()

			io.To(socket.Room(to)).Emit("callEnded")
		})

		// Chat Rooms & Typing
		client.On("joinChat", func(args ...any) {
			if len(args) > 0 {
				if chatID, ok := args[0].(string); ok && chatID != "" {
					client.Join(socket.Room(chatID))
				}
			}
		})

		client.On("leaveChat", func(args ...any) {
			if len(args) > 0 {
				if chatID, ok := args[0].(string); ok && chatID != "" {
					client.Leave(socket.Room(chatID))
				}
			}
		})

		client.On("typing", func(args ...any) {
			if len(args) > 0 {
				if room, ok := args[0].(string); ok {
					client.To(socket.Room(room)).Emit("typing", room)
				}
			}
		})

		client.On("stopTyping", func(args ...any) {
			if len(args) > 0 {
				if room, ok := args[0].(string); ok {
					client.To(socket.Room(room)).Emit("stopTyping", room)
				}
			}
		})

		// disconnect handler
		client.On("disconnect", func(...any) {
			Manager.mu.Lock()
			var disconnectedUserID string
			for uid, sid := range Manager.OnlineUsers {
				if sid == socketID {
					disconnectedUserID = uid
					delete(Manager.OnlineUsers, uid)
					break
				}
			}

			delete(Manager.RingingCalls, socketID)

			for roomID, room := range Manager.VideoRooms {
				for uid, p := range room {
					if p.SocketID == string(socketID) {
						delete(room, uid)
						participants := make([]map[string]string, 0, len(room))
						for _, pItem := range room {
							participants = append(participants, map[string]string{
								"id":   pItem.UserID,
								"name": pItem.UserName,
							})
						}
						if len(room) == 0 {
							delete(Manager.VideoRooms, roomID)
						}
						io.To(socket.Room(roomID)).Emit("userLeftRoom", map[string]any{
							"userId":       uid,
							"participants": participants,
						})
					}
				}
			}

			onlineKeys := make([]string, 0, len(Manager.OnlineUsers))
			for k := range Manager.OnlineUsers {
				onlineKeys = append(onlineKeys, k)
			}
			Manager.mu.Unlock()

			if disconnectedUserID != "" {
				io.Emit("onlineUsers", onlineKeys)
			}
		})
	})

	log.Println("[Socket] Socket.IO v4 handler initialized")
	return io, io.ServeHandler(nil)
}

// BroadcastToRoom emits an event to a specific room
func BroadcastToRoom(room string, event string, data any) {
	if Manager != nil && Manager.Server != nil {
		Manager.Server.To(socket.Room(room)).Emit(event, data)
	}
}

// BroadcastGlobal emits an event to all connected clients
func BroadcastGlobal(event string, data any) {
	if Manager != nil && Manager.Server != nil {
		Manager.Server.Emit(event, data)
	}
}
